### ISSUE TITLE
Support TLS False Start

### ISSUE THREAD

**Issue by scalablecory** [dotnet/runtime#45047]
TLS False Start is a way to reduce the TLS 1.2 handshake by 1-RTT. Doing this would make HttpClient, for example, reduce the latency of its first request to a server by 25%. Combined with [TCP Fast Open](https://github.com/dotnet/runtime/issues/1476), it would reduce that time to 50%.

It is supported by [Schannel](https://docs.microsoft.com/en-us/windows/win32/api/schannel/ns-schannel-secpkgcontext_earlystart) and by [macOS](https://developer.apple.com/documentation/security/2976278-sec_protocol_options_set_tls_fal?language=objc). From what I can tell, not OpenSSL.

One note against this is that TLS 1.3 has this baked in, so it will only be used up to TLS 1.2.

@wfurt

**Comment by geoffkizer** [comment]
https://tools.ietf.org/html/rfc7918

**Comment by geoffkizer** [comment]
I tend to think this isn't that interesting because TLS 1.3 solves the problem. But I'm open to being convinced otherwise.

**Comment by scalablecory** [comment]
TLS 1.2 is going to be around for a while. I think if it's trivial to do, it is a compelling feature.

If it's a large amount of work, then I agree much less interesting.

**Comment by scalablecory** [comment]
Implementing it looks fairly simple, though usage is interesting. The change would look something like:

- AuthenticateAsClient() does 90% of the work it does today, but returns before the server's final frame handshake. Things are still safe (authenticated/etc.) here, but the handshake might fail later.

- Send/Receive works as usual... but you might get an `AuthenticationException` here if the final frame comes rejecting the client's choices. Any data sent is guaranteed to not have been processed.

So, anyone using this would need to be able to retry whatever they were doing on a new connection, but with False Start disabled.

**Comment by wfurt** [comment]
TLS 1.3 sort of does it and it broke bunch of apps on Windows. AFAIK that was reason to support TLS1.3 only via new API to make sure legacy code does not get impacted. We would probably need to have some sign-in mechanism. 
While I agree that TLS 1.2 may be around for a while I would feel that TLS1.3 provides clear path forward and should be available if somebody really cares.

**Comment by geoffkizer** [comment]
> So, anyone using this would need to be able to retry whatever they were doing on a new connection, but with False Start disabled

That seems to imply that False Start should be opt-in. Is that correct?

Given that, how common do we think it would be to opt-in? e.g. would we do it in SocketsHttpHandler?

**Comment by scalablecory** [comment]
> would we do it in SocketsHttpHandler?

For `SocketsHttpHandler` specifically:

In terms of failure mode it looks very rare, so I think there's little downside to enabling it by default. Basically only if server is doing something weird, but we can reach out to SChannel team to get more details.

Complexity-wise, I don't think it's huge, but still might be too much to be worth the effort. I see high level as this:
- Add a `bool HttpContent.CanResend { get; }`, maybe just internal for our own types.
- Add a `Task SslStream.HandshakeComplete { get; }` that completes once the final leg of handshake is done.
- When opening a new connection, disable False Start if `CanResend == false` or if the `Stream` is not an `SslStream`. (the latter part would not be needed if we do System.Net.Connections)
- When opening a new HTTP/2 connection, defer sending any requests with `CanResend == false` until `HandshakeComplete` finishes.
- If `AuthenticationException` is thrown during `Read`/`Write`, restart the request without False Start.

**Comment by scalablecory** [comment]
Closing in favor of #49274.
